deduce.go

```diff
@@ -690,6 +695,253 @@ func (sm *SourceMgr) deduceFromPath(path string) (deductionFuture, error) {
 	}, nil
 }
 
+type deductionCoordinator struct {
+	ctx      context.Context
```

Comment:
I think vanilla contexts handle that pretty well with just one tree of contexts, I don't see how you need to combine cancelation paths. I imagine it looking something like this:

```go
func main() {
  ctx, cancel := context.WithCancel(context.Background())
  go cancelOnSignal(cancel) // yknow, set up a loop on a chan signal.Signal, blah blah

  sm := NewSourceManager(ctx, [...]) 
}
```

If the root context is canceled because of a signal, that gets propagated down to all the children, which will be all the open calls, like you want.

For the second case, if you want a 'locally' cancelable function, you call `context.WithCancel` again. The newly generated context will cancel if the old papa context made in `main` gets canceled, or if the new local `cancel` method is called. But the local cancel method won't close any of the context's ancestors. This seems to be what you want, right?